File: lib/features/activity_freediving/models/freediving_details.dart
enum WaterBody { sea, fjord, lake }
enum BottomType { unknown, sandyShallow, rockyShallow, kelpForest, wall, reef, seagrassMeadow, open }

class FreedivingDetails {
  final WaterBody waterBody;
  final BottomType bottomType;
  final double maxDepthMeters;
  final double? typicalVisibilityMeters;
  final bool harpoonAllowed;
  final bool shoreEntry;
  final String? accessNotes;
  final List<TargetSpecies> targetSpecies;

  const FreedivingDetails({
    required this.waterBody,
    required this.bottomType,
    required this.maxDepthMeters,
    this.typicalVisibilityMeters,
    this.harpoonAllowed = false,
    this.shoreEntry = true,
    this.accessNotes,
    this.targetSpecies = const [],
  });

  Map<String, dynamic> toJson() => {
        'waterBody': waterBody.index,
        'bottomType': bottomType.index,
        'maxDepthMeters': maxDepthMeters,
        'typicalVisibilityMeters': ?typicalVisibilityMeters,
        'harpoonAllowed': harpoonAllowed,
        'shoreEntry': shoreEntry,
        'accessNotes': ?accessNotes,
        'targetSpecies': targetSpecies.map((t) => t.toJson()).toList(),
      };

  factory FreedivingDetails.fromJson(Map<String, dynamic> json) => FreedivingDetails(
        waterBody: WaterBody.values[(json['waterBody'] as num).toInt()],
        bottomType: BottomType.values[(json['bottomType'] as num).toInt()],
        maxDepthMeters: (json['maxDepthMeters'] as num).toDouble(),
        typicalVisibilityMeters: (json['typicalVisibilityMeters'] as num?)?.toDouble(),
        harpoonAllowed: json['harpoonAllowed'] as bool? ?? false,
        shoreEntry: json['shoreEntry'] as bool? ?? true,
        accessNotes: json['accessNotes'] as String?,
        targetSpecies: (json['targetSpecies'] as List?)
                ?.cast<Map<String, dynamic>>()
                .map(TargetSpecies.fromJson)
                .toList() ?? const [],
      );
}

class TargetSpecies {
  final String speciesCode;
  final String? notes;
  const TargetSpecies({required this.speciesCode, this.notes});

  Map<String, dynamic> toJson() => {'speciesCode': speciesCode, 'notes': ?notes};
  factory TargetSpecies.fromJson(Map<String, dynamic> json) => TargetSpecies(
        speciesCode: json['speciesCode'] as String,
        notes: json['notes'] as String?,
      );
}

File: apps/api/src/Activities/Turbo.Activities.Freediving.Contracts/value/FreedivingDetails.cs
using System.Text.Json.Serialization;

namespace Turboapi.Activities.Freediving.value;

public sealed record FreedivingDetails
{
    [JsonPropertyName("waterBody")] public WaterBody WaterBody { get; init; }
    [JsonPropertyName("bottomType")] public BottomType BottomType { get; init; }
    [JsonPropertyName("maxDepthMeters")] public float MaxDepthMeters { get; init; }
    [JsonPropertyName("typicalVisibilityMeters")] public float? TypicalVisibilityMeters { get; init; }
    [JsonPropertyName("harpoonAllowed")] public bool HarpoonAllowed { get; init; }
    [JsonPropertyName("shoreEntry")] public bool ShoreEntry { get; init; }
    [JsonPropertyName("accessNotes")] public string? AccessNotes { get; init; }
    [JsonPropertyName("targetSpecies")] public IReadOnlyList<TargetSpecies> TargetSpecies { get; init; } = Array.Empty<TargetSpecies>();

    [JsonConstructor]
    public FreedivingDetails(
        WaterBody waterBody, BottomType bottomType, float maxDepthMeters,
        float? typicalVisibilityMeters, bool harpoonAllowed, bool shoreEntry,
        string? accessNotes, IReadOnlyList<TargetSpecies>? targetSpecies)
    {
        WaterBody = waterBody;
        BottomType = bottomType;
        MaxDepthMeters = maxDepthMeters;
        TypicalVisibilityMeters = typicalVisibilityMeters;
        HarpoonAllowed = harpoonAllowed;
        ShoreEntry = shoreEntry;
        AccessNotes = accessNotes;
        TargetSpecies = targetSpecies ?? Array.Empty<TargetSpecies>();
    }
}

public sealed record TargetSpecies
{
    [JsonPropertyName("speciesCode")] public string SpeciesCode { get; init; }
    [JsonPropertyName("notes")] public string? Notes { get; init; }

    [JsonConstructor]
    public TargetSpecies(string speciesCode, string? notes)
    {
        SpeciesCode = speciesCode ?? throw new ArgumentNullException(nameof(speciesCode));
        Notes = notes;
    }
}
